### PR TITLE
BCDA-4234 - Add functionality to encode models.Alr to FHIR resources.

### PR DESCRIPTION
**Pull request by mtrang1263**
### Fixes [BCDA-4234](https://jira.cms.gov/browse/BCDA-4234)

Using the strategy documented here, we add a new ALR package that will translate the models.Alr.

### Change Details

1. Adding new alr package underneath models/fhir that contains all of the logic that translates the models.Alr to Patient & Observation resources.
2. Moved out some worker specific code out of models.go into the bcdaworker. This was required because of the import cycle that occurred when we added the fhir/alr package.
3. Slight tweaks on the alr generation/deserializer based on the usages in the fhir/alr
4. Adding test that can produce random ALR data encoded as Patient + Observation.

### Security Implications

- [ ] new software dependencies
- [ ] security controls or supporting software altered
- [x] new data stored or transmitted

Not PHI/PII related, but we are storing the HCC crosswalk data. This is required for us to translate the ALR data to something usable by the end user.

- [ ] security checklist is completed for this change
- [ ] requires more information or team discussion to evaluate security implications
- [x] no PHI/PII is affected by this change

### Generated files
[alr_fhir_gen.zip](https://github.com/CMSgov/bcda-app/files/6077614/alr_fhir_gen.zip)

### Acceptance Validation
1. Deploy to CI to ensure that build & package places the crosswalk file in the expected.
2. Review produced ALR FHIR resources with Dan and Bill.

### CODE REVIEW

File: bcda/alr/csv/csv.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ var joinFields = requiredFields
 
 // ToALR reads in a CSV file(s) and unmarshals the data into an ALR model.
 // CSV files are joined based on a predetermined list of fields
-func ToALR(csvPaths ...string) ([]models.Alr, error) {
+func ToALR(csvPaths ...string) ([]*models.Alr, error) {
```
Comment (CosmicSyntax) [inline bcda/alr/csv/csv.go:38]: 👍

File: bcda/alr/gen/synthetic.go [hunk 2]
```diff
@@ -66,6 +66,12 @@ var valuegen map[*regexp.Regexp]func() string = map[*regexp.Regexp]func() string
 			return strconv.FormatFloat(res, 'f', 3, 64)
 		})
 	},
+	regexp.MustCompile("_EXCLUDED$"): func() string {
+		return randomEmpty(half,
+			func() string { return strconv.Itoa(randomdata.Number(2)) })
+	},
+	regexp.MustCompile("^CBA_FLAG$"):        func() string { return strconv.Itoa(randomdata.Number(2)) },
+	regexp.MustCompile("^ASSIGNMENT_TYPE$"): func() string { return strconv.Itoa(randomdata.Number(3)) },
```
Comment (CosmicSyntax) [inline bcda/alr/gen/synthetic.go:74]: Is there a document that goes into these fields?

File: bcda/alr/gen/synthetic.go [hunk 3]
```diff
@@ -66,6 +66,12 @@ var valuegen map[*regexp.Regexp]func() string = map[*regexp.Regexp]func() string
 			return strconv.FormatFloat(res, 'f', 3, 64)
 		})
 	},
+	regexp.MustCompile("_EXCLUDED$"): func() string {
```
Comment (CosmicSyntax) [inline bcda/alr/gen/synthetic.go:69]: Can we change this to "EXCLUDED$" to capture this too:
Screen Shot 2021-03-03 at 6 59 45 PM

File: bcda/models/fhir/alr/observation.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+			},
+			{System: fhirURI("http://cms.gov/CodeSystem/alr"),
+				Code:    &fhirdatatypes.Code{Value: code},
+				Display: fhirString(display),
+			},
+		},
+	}
+}
+
+func observationComponents(pattern *regexp.Regexp, system string, keyValue map[string]string) []*fhirmodels.Observation_Component {
```
Comment (CosmicSyntax) [inline bcda/models/fhir/alr/observation.go:108]: Is pattern going to be used later on?

File: ops/build_and_package.sh [hunk 5]
```diff
@@ -35,12 +35,18 @@ then
   exit 1
 fi
 
+if [ ! -f ../bcda/models/fhir/alr/hcc_crosswalk.csv ]
+then
+  echo "Crosswalk flie must exist prior to creating package."
```
Comment (CosmicSyntax) [inline ops/build_and_package.sh:40]: file?

File: ops/build_and_package.sh [hunk 6]
```diff
@@ -35,12 +35,18 @@ then
...
+  exit 1
+fi
+
 cd ../bcda
 go clean
 echo "Building bcda binary..." 
 go build -ldflags "-X github.com/CMSgov/bcda-app/bcda/constants.Version=$VERSION"
 echo "Packaging bcda binary into RPM..."
-fpm -v $VERSION -s dir -t rpm -n bcda bcda=/usr/local/bin/bcda swaggerui=/etc/sv/api
+fpm -v $VERSION -s dir -t rpm -n bcda bcda=/usr/local/bin/bcda swaggerui=/etc/sv/api models/fhir/alr/hcc_crosswalk.csv=/etc/sv/api/hcc_crosswalk.csv
```
Comment (CosmicSyntax) [inline ops/build_and_package.sh:49]: I wonder if we should have the conf package handle the crosswalk.

File: bcda/models/fhir/alr/patient.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	p.Identifier = getIdentifiers(alr)
+	return p
+}
+
+func getGender(gender string) *fhircodes.AdministrativeGenderCode {
+	if gender == "1" {
+		return &fhircodes.AdministrativeGenderCode{Value: fhircodes.AdministrativeGenderCode_MALE}
+	} else if gender == "2" {
+		return &fhircodes.AdministrativeGenderCode{Value: fhircodes.AdministrativeGenderCode_FEMALE}
+	} else {
```
Comment (CosmicSyntax) [inline bcda/models/fhir/alr/patient.go:29]: Should we include "0" for Unknown / AdministrativeGenderCode_UNKNOWN?

File: bcda/models/fhir/alr/observation.go (hunk 4 above)
Comment (mtrang1263) [reply bcda/models/fhir/alr/observation.go:108]: I'm thinking this gives us the flexibility to encode N number of observation resources. What're your thoughts on it?

File: ops/build_and_package.sh (hunk 6 above)
Comment (mtrang1263) [reply ops/build_and_package.sh:49]: Let's chat about this when you're online. I was wondering if this falls in as an application config or application data. Thinking more it as data that is stored as a file. But definitely open to see if it makes more sense in conf package/stored elsewhere.

**Review by CosmicSyntax — APPROVED** [review]
LGTM

**Pull request merged**